group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java

```diff
@@ -0,0 +1,959 @@
+         */
+        final TopicPartition tp;
+
+        /**
+         * The operation name.
+         */
+        final String name;
+
+        /**
+         * The write operation to execute.
+         */
+        final CoordinatorWriteOperation<S, T, U> op;
+
+        /**
+         * The future that will be completed with the response
+         * generated by the write operation or an error.
+         */
+        final CompletableFuture<T> future;
+
+        /**
+         * The result of the write operation. It could be null
+         * if an exception is thrown before it is assigned.
+         */
+        CoordinatorResult<T, U> result;
+
+        /**
+         * Constructor.
+         *
+         * @param name  The operation name.
+         * @param tp    The topic partition that the operation is applied to.
+         * @param op    The write operation.
+         */
+        CoordinatorWriteEvent(
+            String name,
+            TopicPartition tp,
+            CoordinatorWriteOperation<S, T, U> op
+        ) {
+            this.tp = tp;
+            this.name = name;
+            this.op = op;
+            this.future = new CompletableFuture<>();
+        }
+
+        /**
+         * @return The key used by the CoordinatorEventProcessor to ensure
+         * that events with the same key are not processed concurrently.
+         */
+        @Override
+        public TopicPartition key() {
+            return tp;
+        }
+
+        /**
+         * Called by the CoordinatorEventProcessor when the event is executed.
+         */
+        @Override
+        public void run() {
+            try {
+                // Get the context of the coordinator or fail if the coordinator is not in active state.
+                CoordinatorContext context = activeContextOrThrow(tp);
+                long prevLastWrittenOffset = context.lastWrittenOffset;
+
+                // Execute the operation.
+                result = op.generateRecordsAndResult(context.coordinator);
+
+                if (result.records().isEmpty()) {
+                    // If the records are empty, it was a read operation after all. In this case,
+                    // the response can be returned directly iff there are no pending write operations;
+                    // otherwise, the read needs to wait on the last write operation to be completed.
+                    OptionalLong pendingOffset = context.deferredEventQueue.highestPendingOffset();
+                    if (pendingOffset.isPresent()) {
+                        context.deferredEventQueue.add(pendingOffset.getAsLong(), this);
+                    } else {
+                        complete(null);
+                    }
+                } else {
+                    // If the records are not empty, first, they are applied to the state machine,
+                    // second, then are written to the partition/log, and finally, the response
+                    // is put into the deferred event queue.
```

Comment:
we can apply to the state machine first because we will revert to latest committed snapshot if the append fails right?

Reply:
Correct.